### ISSUE TITLE
Test hooks in modular suite are not called in an expected order

### ISSUE THREAD

**Issue by odockal** [vitest-dev/vitest#3652]
### Describe the bug

I believe, that a way how the hooks are being resolved in the test suite is not right `*`. 

I am trying to achieve a mocha(junit)-like behavior when defining my testing scenarios. Having a suite spec file that defines a setup/teardown behavior for all the test cases inside of it, and I would also like to have the same option to (call before/after[all/each]) in each particular test case/file. 

```
// suite.spec.ts
import { testFeatureA } from 'testFeatureA.ts'
...
describe('Smoke test suite', async () => {
    beforeAll(async () => {...});
    afterAll(async () => {...});
    await testFeatureA();
    await testFeatureB();
})
```
```
// testFeatureA.ts
export async function testFeatureA() {
    describe(`${name} tests`, async () => {
        beforeAll( async () => {...});
        afterAll( async () => {...});
        test(`test number 1`, async () => {...});
```
Having this results in:
```
Before All smoke suite 
Before All test FeatureA
...
Before Each test FeatureA
test number 1
After Each test FeatureA
Before Each test FeatureB
test number 1
After Each test FeatureB
...
After All test FeatureA --> wrong
Before All test FeatureB --> wrong
After All test FeatureB
After All smoke suite
```

`*` I cannot really exclude an option that I am not doing it right as well

### Reproduction

I have created a [demo repo] which should easily point out the problem and how to reproduce it.

[demo repo]: https://github.com/odockal/vitest-hierarchy

### System Info

```shell
System:
    OS: Linux 6.2 Fedora Linux 37 (Workstation Edition)
    CPU: (12) x64 Intel(R) Core(TM) i7-10850H CPU @ 2.70GHz
    Memory: 16.10 GB / 31.09 GB
    Container: Yes
    Shell: 5.9 - /usr/bin/zsh
  Binaries:
    Node: 18.15.0 - /usr/bin/node
    Yarn: 1.22.19 - /usr/bin/yarn
    npm: 9.5.0 - /usr/bin/npm
  Browsers:
    Chrome: 113.0.5672.63
    Firefox: 112.0.2
  npmPackages:
    vitest: ^0.32.2 => 0.32.2
```

### Used Package Manager

npm

### Validations

- [X] Follow our [Code of Conduct](https://github.com/vitest-dev/vitest/blob/main/CODE_OF_CONDUCT.md)
- [X] Read the [Contributing Guidelines](https://github.com/vitest-dev/vitest/blob/main/CONTRIBUTING.md).
- [X] Read the [docs](https://vitest.dev/guide/).
- [X] Check that there isn't [already an issue](https://github.com/vitest-dev/vitest/issues) that reports the same bug to avoid creating a duplicate.
- [X] Check that this is a concrete bug. For Q&A open a [GitHub Discussion](https://github.com/vitest-dev/vitest/discussions) or join our [Discord Chat Server](https://chat.vitest.dev).
- [X] The provided reproduction is a [minimal reproducible example](https://stackoverflow.com/help/minimal-reproducible-example) of the bug.

**Comment by sheremet-va** [comment]
You can change the order of hook execution with [`sequence.hooks`](https://vitest.dev/config/#sequence-hooks) option

**Comment by odockal** [comment]
@sheremet-va Thanks for a hint, I tried but did not really worked. Neither of `stack`, `list` or `parallel` helped. It seems wrong that I got particular `before/afterAll` called outside the context of their definition (exported through function in describe block) expect first and global all hook.

**Comment by sheremet-va** [comment]
I don't think this was ever an actual issue. The reason why you see logs in this order is because Vitest batches them and prints under the same `unknown test` label. This batching logic was fixed at one point, so if you try the reproduction on the latest version, it works correctly (it also _worked_ correctly before in the sense that callbacks were executed in the right order).
